gadget/gadget.go

```diff
@@ -273,15 +282,15 @@ func systemOrSnapID(s string) bool {
 	return true
 }
 
-// ReadInfo reads the gadget specific metadata from gadget.yaml
-// in the snap. classic set to true means classic rules apply,
-// i.e. content/presence of gadget.yaml is fully optional.
-func ReadInfo(gadgetSnapRootDir string, classic bool) (*Info, error) {
+// ReadInfo reads the gadget specific metadata from gadget.yaml in the snap. If
+// constraints is nil, ReadInfo will just check for self-consistency, otherwise
+// rules for the classic or system seed cases are enforced.
+func ReadInfo(gadgetSnapRootDir string, constraints *ModelConstraints) (*Info, error) {
 	var gi Info
 
 	gadgetYamlFn := filepath.Join(gadgetSnapRootDir, "meta", "gadget.yaml")
 	gmeta, err := ioutil.ReadFile(gadgetYamlFn)
-	if classic && os.IsNotExist(err) {
+	if constraints != nil && constraints.Classic && os.IsNotExist(err) {
```

Comment:
if constraints is nil we don't know,  so it's really `(constraints == nil || constraints.Classic) && os.IsNoExistst` I think

Reply:
And if we don't have the gadget yaml there will be no consistency to be checked so yes, we don't know at this point and it should be as you described.